src/Thentos/Api.hs

```diff
@@ -378,21 +393,34 @@ _sessionAndUserIdFromToken tok = do
         UserA uid -> return (session, uid)
         ServiceA _ -> lift $ left OperationNotPossibleInServiceSession
 
--- | FIXME: 'addServiceLogin' and 'dropServiceLogin' still have the
--- old 'is logged in'-semantics.  we need to switch to the new
--- semantics where the transactions used here register users with
--- services, and there need to be separate transactions for login,
--- logout.
-addServiceLogin :: SessionToken -> ServiceId -> Action r ()
-addServiceLogin tok sid = do
+addServiceRegistration :: SessionToken -> ServiceId -> Action r ()
+addServiceRegistration tok sid = do
     (_, uid) <- _sessionAndUserIdFromToken tok
-    updateAction $ UpdateUserField uid (UpdateUserFieldAddService sid)
+    updateAction $ UpdateUserField uid (UpdateUserFieldAddService sid newServiceAccount)
 
-dropServiceLogin :: SessionToken -> ServiceId -> Action r ()
-dropServiceLogin tok sid = do
+dropServiceRegistration :: SessionToken -> ServiceId -> Action r ()
+dropServiceRegistration tok sid = do
     (_, uid) <- _sessionAndUserIdFromToken tok
     updateAction $ UpdateUserField uid (UpdateUserFieldDropService sid)
 
+-- | Registers user implicitly if not registered yet.
+updateServiceRegistration :: (ServiceAccount -> ServiceAccount)
+        -> SessionToken -> ServiceId -> Action r ()
+updateServiceRegistration f tok sid = do
+    (_, uid) <- _sessionAndUserIdFromToken tok
+    (_, user) <- queryAction (LookupUser uid)
+    let account = fromMaybe newServiceAccount . Map.lookup sid $ user ^. userServices
+    updateAction $ UpdateUserField uid (UpdateUserFieldAddService sid (f account))
+
+-- | Registers user implicitly if not registered yet.
+addServiceLogin :: SessionToken -> ServiceId -> Action r ()
+addServiceLogin = updateServiceRegistration $ serviceSessionTimeout .~ Just defaultSessionTimeout
+
+-- | If user is not registered with service, do nothing.  Registers
+-- user implicitly if not registered yet.
```

Comment:
I should have seen this earlier, but these two sentences contradict each other, don't they?